File: c/src/keymap.c
/*
 * keymap.c — HID keycode mappings and typo neighbor keys
 */
#include "keymap.h"

#include <string.h>
#include <stdlib.h>
#include <ctype.h>

/* ── HID Usage IDs ─────────────────────────────────────────────────────────── */
/* Each entry: { modifier, keycode } */

typedef struct {
    char ch;
    uint8_t modifier;
    uint8_t keycode;
} keymap_entry_t;

static const keymap_entry_t KEYMAP[] = {
    /* Whitespace */
    { ' ',  0x00, 0x2C }, { '\n', 0x00, 0x28 }, { '\t', 0x00, 0x2B },

    /* Lowercase letters */
    { 'a', 0x00, 0x04 }, { 'b', 0x00, 0x05 }, { 'c', 0x00, 0x06 },
    { 'd', 0x00, 0x07 }, { 'e', 0x00, 0x08 }, { 'f', 0x00, 0x09 },
    { 'g', 0x00, 0x0A }, { 'h', 0x00, 0x0B }, { 'i', 0x00, 0x0C },
    { 'j', 0x00, 0x0D }, { 'k', 0x00, 0x0E }, { 'l', 0x00, 0x0F },
    { 'm', 0x00, 0x10 }, { 'n', 0x00, 0x11 }, { 'o', 0x00, 0x12 },
    { 'p', 0x00, 0x13 }, { 'q', 0x00, 0x14 }, { 'r', 0x00, 0x15 },
    { 's', 0x00, 0x16 }, { 't', 0x00, 0x17 }, { 'u', 0x00, 0x18 },
    { 'v', 0x00, 0x19 }, { 'w', 0x00, 0x1A }, { 'x', 0x00, 0x1B },
    { 'y', 0x00, 0x1C }, { 'z', 0x00, 0x1D },

    /* Uppercase letters (Left Shift = 0x02) */
    { 'A', 0x02, 0x04 }, { 'B', 0x02, 0x05 }, { 'C', 0x02, 0x06 },
    { 'D', 0x02, 0x07 }, { 'E', 0x02, 0x08 }, { 'F', 0x02, 0x09 },
    { 'G', 0x02, 0x0A }, { 'H', 0x02, 0x0B }, { 'I', 0x02, 0x0C },
    { 'J', 0x02, 0x0D }, { 'K', 0x02, 0x0E }, { 'L', 0x02, 0x0F },
    { 'M', 0x02, 0x10 }, { 'N', 0x02, 0x11 }, { 'O', 0x02, 0x12 },
    { 'P', 0x02, 0x13 }, { 'Q', 0x02, 0x14 }, { 'R', 0x02, 0x15 },
    { 'S', 0x02, 0x16 }, { 'T', 0x02, 0x17 }, { 'U', 0x02, 0x18 },
    { 'V', 0x02, 0x19 }, { 'W', 0x02, 0x1A }, { 'X', 0x02, 0x1B },
    { 'Y', 0x02, 0x1C }, { 'Z', 0x02, 0x1D },

    /* Numbers */
    { '1', 0x00, 0x1E }, { '2', 0x00, 0x1F }, { '3', 0x00, 0x20 },
    { '4', 0x00, 0x21 }, { '5', 0x00, 0x22 }, { '6', 0x00, 0x23 },
    { '7', 0x00, 0x24 }, { '8', 0x00, 0x25 }, { '9', 0x00, 0x26 },
    { '0', 0x00, 0x27 },

    /* Shifted number symbols */
    { '!', 0x02, 0x1E }, { '@', 0x02, 0x1F }, { '#', 0x02, 0x20 },
    { '$', 0x02, 0x21 }, { '%', 0x02, 0x22 }, { '^', 0x02, 0x23 },
    { '&', 0x02, 0x24 }, { '*', 0x02, 0x25 }, { '(', 0x02, 0x26 },
    { ')', 0x02, 0x27 },

    /* Punctuation */
    { '-', 0x00, 0x2D }, { '_', 0x02, 0x2D }, { '=', 0x00, 0x2E },
    { '+', 0x02, 0x2E }, { '[', 0x00, 0x2F }, { '{', 0x02, 0x2F },
    { ']', 0x00, 0x30 }, { '}', 0x02, 0x30 }, { '\\', 0x00, 0x31 },
    { '|', 0x02, 0x31 }, { ';', 0x00, 0x33 }, { ':', 0x02, 0x33 },
    { '\'', 0x00, 0x34 }, { '"', 0x02, 0x34 }, { '`', 0x00, 0x35 },
    { '~', 0x02, 0x35 }, { ',', 0x00, 0x36 }, { '<', 0x02, 0x36 },
    { '.', 0x00, 0x37 }, { '>', 0x02, 0x37 }, { '/', 0x00, 0x38 },
    { '?', 0x02, 0x38 },
};

#define KEYMAP_SIZE (sizeof(KEYMAP) / sizeof(KEYMAP[0]))

int char_to_hid_report(char c, uint8_t report[8]) {
    for (size_t i = 0; i < KEYMAP_SIZE; i++) {
        if (KEYMAP[i].ch == c) {
            report[0] = 0x01;              /* Report ID (keyboard) */
            report[1] = KEYMAP[i].modifier;
            report[2] = 0x00;              /* Reserved */
            report[3] = KEYMAP[i].keycode;
            report[4] = 0x00;
            report[5] = 0x00;
            report[6] = 0x00;
            report[7] = 0x00;
            return 0;
        }
    }
    return -1;
}

void release_report(uint8_t report[8]) {
    report[0] = 0x01;  /* Report ID (keyboard) */
    memset(report + 1, 0, 7);
}

/* ── Nearby keys for typo simulation ──────────────────────────────────────── */

typedef struct {
    char key;
    const char *neighbors;
} neighbor_entry_t;

static const neighbor_entry_t NEARBY_KEYS[] = {
    { 'a', "sqwz"   }, { 'b', "vghn"   }, { 'c', "xdfv"   },
    { 'd', "sfgxce" }, { 'e', "wsdr"   }, { 'f', "dgrtvc" },
    { 'g', "fhtybv" }, { 'h', "gjyun"  }, { 'i', "uojk"   },
    { 'j', "hkuim"  }, { 'k', "jloi"   }, { 'l', "kop"    },
    { 'm', "njk"    }, { 'n', "bhmj"   }, { 'o', "iklp"   },
    { 'p', "ol"     }, { 'q', "wa"     }, { 'r', "etfd"   },
    { 's', "adwxze" }, { 't', "ryfg"   }, { 'u', "yhji"   },
    { 'v', "cfgb"   }, { 'w', "qase"   }, { 'x', "zsdc"   },
    { 'y', "tghu"   }, { 'z', "asx"    },
};

#define NEARBY_SIZE (sizeof(NEARBY_KEYS) / sizeof(NEARBY_KEYS[0]))

char get_neighbor_key(char c) {
    char lower = (char)tolower((unsigned char)c);
    for (size_t i = 0; i < NEARBY_SIZE; i++) {
        if (NEARBY_KEYS[i].key == lower) {
            const char *n = NEARBY_KEYS[i].neighbors;
            size_t len = strlen(n);
            if (len == 0) return c;
            return n[rand() % len];
        }
    }
    return c;
}

File: rust/src/keymap.rs
use std::collections::HashMap;
use std::sync::LazyLock;

/// USB HID keycode entry: (modifier_byte, keycode)
#[derive(Debug, Clone, Copy)]
pub struct HidKey {
    pub modifier: u8,
    pub keycode: u8,
}

pub const BACKSPACE: HidKey = HidKey {
    modifier: 0x00,
    keycode: 0x2A,
};

/// Build an 8-byte keyboard report (report_id=0x01).
/// Format: [report_id, modifier, 0x00, keycode, 0, 0, 0, 0, 0]
/// We return 9 bytes (including report ID) to match the Python impl.
pub fn char_to_hid_report(c: char) -> Option<[u8; 9]> {
    let key = KEYMAP.get(&c)?;
    Some(make_key_report(key.modifier, key.keycode))
}

pub fn make_key_report(modifier: u8, keycode: u8) -> [u8; 9] {
    [0x01, modifier, 0x00, keycode, 0, 0, 0, 0, 0]
}

/// All-zeros key release report (report_id=0x01).
pub fn key_release_report() -> [u8; 9] {
    [0x01, 0, 0, 0, 0, 0, 0, 0, 0]
}

/// Build a 5-byte mouse report (report_id=0x02).
/// Format: [report_id, buttons, dx, dy, wheel]
pub fn mouse_report(buttons: u8, dx: i8, dy: i8, wheel: i8) -> [u8; 5] {
    [0x02, buttons, dx as u8, dy as u8, wheel as u8]
}

/// Nearby keys on QWERTY for typo simulation.
pub static NEARBY_KEYS: LazyLock<HashMap<char, &'static str>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    m.insert('a', "sqwz");
    m.insert('b', "vghn");
    m.insert('c', "xdfv");
    m.insert('d', "sfgxce");
    m.insert('e', "wsdr");
    m.insert('f', "dgrtvc");
    m.insert('g', "fhtybv");
    m.insert('h', "gjyun");
    m.insert('i', "uojk");
    m.insert('j', "hkuim");
    m.insert('k', "jloi");
    m.insert('l', "kop");
    m.insert('m', "njk");
    m.insert('n', "bhmj");
    m.insert('o', "iklp");
    m.insert('p', "ol");
    m.insert('q', "wa");
    m.insert('r', "etfd");
    m.insert('s', "adwxze");
    m.insert('t', "ryfg");
    m.insert('u', "yhji");
    m.insert('v', "cfgb");
    m.insert('w', "qase");
    m.insert('x', "zsdc");
    m.insert('y', "tghu");
    m.insert('z', "asx");
    m
});

/// Full USB HID keycode map matching the Python implementation.
static KEYMAP: LazyLock<HashMap<char, HidKey>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    let k = |modifier: u8, keycode: u8| HidKey { modifier, keycode };

    // Whitespace
    m.insert(' ', k(0x00, 0x2C));
    m.insert('\n', k(0x00, 0x28));
    m.insert('\t', k(0x00, 0x2B));

    // Lowercase
    m.insert('a', k(0x00, 0x04));
    m.insert('b', k(0x00, 0x05));
    m.insert('c', k(0x00, 0x06));
    m.insert('d', k(0x00, 0x07));
    m.insert('e', k(0x00, 0x08));
    m.insert('f', k(0x00, 0x09));
    m.insert('g', k(0x00, 0x0A));
    m.insert('h', k(0x00, 0x0B));
    m.insert('i', k(0x00, 0x0C));
    m.insert('j', k(0x00, 0x0D));
    m.insert('k', k(0x00, 0x0E));
    m.insert('l', k(0x00, 0x0F));
    m.insert('m', k(0x00, 0x10));
    m.insert('n', k(0x00, 0x11));
    m.insert('o', k(0x00, 0x12));
    m.insert('p', k(0x00, 0x13));
    m.insert('q', k(0x00, 0x14));
    m.insert('r', k(0x00, 0x15));
    m.insert('s', k(0x00, 0x16));
    m.insert('t', k(0x00, 0x17));
    m.insert('u', k(0x00, 0x18));
    m.insert('v', k(0x00, 0x19));
    m.insert('w', k(0x00, 0x1A));
    m.insert('x', k(0x00, 0x1B));
    m.insert('y', k(0x00, 0x1C));
    m.insert('z', k(0x00, 0x1D));

    // Uppercase (shift = 0x02)
    m.insert('A', k(0x02, 0x04));
    m.insert('B', k(0x02, 0x05));
    m.insert('C', k(0x02, 0x06));
    m.insert('D', k(0x02, 0x07));
    m.insert('E', k(0x02, 0x08));
    m.insert('F', k(0x02, 0x09));
    m.insert('G', k(0x02, 0x0A));
    m.insert('H', k(0x02, 0x0B));
    m.insert('I', k(0x02, 0x0C));
    m.insert('J', k(0x02, 0x0D));
    m.insert('K', k(0x02, 0x0E));
    m.insert('L', k(0x02, 0x0F));
    m.insert('M', k(0x02, 0x10));
    m.insert('N', k(0x02, 0x11));
    m.insert('O', k(0x02, 0x12));
    m.insert('P', k(0x02, 0x13));
    m.insert('Q', k(0x02, 0x14));
    m.insert('R', k(0x02, 0x15));
    m.insert('S', k(0x02, 0x16));
    m.insert('T', k(0x02, 0x17));
    m.insert('U', k(0x02, 0x18));
    m.insert('V', k(0x02, 0x19));
    m.insert('W', k(0x02, 0x1A));
    m.insert('X', k(0x02, 0x1B));
    m.insert('Y', k(0x02, 0x1C));
    m.insert('Z', k(0x02, 0x1D));

    // Numbers
    m.insert('1', k(0x00, 0x1E));
    m.insert('2', k(0x00, 0x1F));
    m.insert('3', k(0x00, 0x20));
    m.insert('4', k(0x00, 0x21));
    m.insert('5', k(0x00, 0x22));
    m.insert('6', k(0x00, 0x23));
    m.insert('7', k(0x00, 0x24));
    m.insert('8', k(0x00, 0x25));
    m.insert('9', k(0x00, 0x26));
    m.insert('0', k(0x00, 0x27));

    // Shifted symbols
    m.insert('!', k(0x02, 0x1E));
    m.insert('@', k(0x02, 0x1F));
    m.insert('#', k(0x02, 0x20));
    m.insert('$', k(0x02, 0x21));
    m.insert('%', k(0x02, 0x22));
    m.insert('^', k(0x02, 0x23));
    m.insert('&', k(0x02, 0x24));
    m.insert('*', k(0x02, 0x25));
    m.insert('(', k(0x02, 0x26));
    m.insert(')', k(0x02, 0x27));

    // Punctuation
    m.insert('-', k(0x00, 0x2D));
    m.insert('_', k(0x02, 0x2D));
    m.insert('=', k(0x00, 0x2E));
    m.insert('+', k(0x02, 0x2E));
    m.insert('[', k(0x00, 0x2F));
    m.insert('{', k(0x02, 0x2F));
    m.insert(']', k(0x00, 0x30));
    m.insert('}', k(0x02, 0x30));
    m.insert('\\', k(0x00, 0x31));
    m.insert('|', k(0x02, 0x31));
    m.insert(';', k(0x00, 0x33));
    m.insert(':', k(0x02, 0x33));
    m.insert('\'', k(0x00, 0x34));
    m.insert('"', k(0x02, 0x34));
    m.insert('`', k(0x00, 0x35));
    m.insert('~', k(0x02, 0x35));
    m.insert(',', k(0x00, 0x36));
    m.insert('<', k(0x02, 0x36));
    m.insert('.', k(0x00, 0x37));
    m.insert('>', k(0x02, 0x37));
    m.insert('/', k(0x00, 0x38));
    m.insert('?', k(0x02, 0x38));

    m
});
